middle_end/closure/closure.ml

```diff
@@ -884,8 +888,19 @@ let close_approx_var { fenv; cenv } id =
   match approx with
     Value_const c -> make_const c
   | approx ->
-      let subst = try V.Map.find id cenv with Not_found -> Uvar id in
-      (subst, approx)
+      match cenv with
+      | Not_in_closure -> Uvar id, approx
+      | In_closure { entries; env_param; env_pos } ->
+        let subst =
+          match V.Map.find id entries with
+          | Free_variable fv_pos ->
+            Uprim(P.Pfield(fv_pos - env_pos, Pointer, Immutable),
+                  [Uvar env_param], Debuginfo.none)
```

Comment:
Before this code fragment would be computed once per function and free variable, and shared between all occurrences of the free variable. Now it is regenerated afresh on each occurrence. I wonder if this could lead to a noticeable loss of sharing / increase in memory usage.

I can think of several ways of sharing these accessor code fragments, including some that would let you share the code of different free variables with the same `fv_pos - env_pos` relative offset. I am not sure which one would be simple enough.

Reply:
I've thought about preserving sharing, but in the end I estimated that it wasn't worth it (keep in mind that even before this PR, the sharing would be lost when we translate to Cmm).

You could try to construct a worst case example and see the impact it has (something like `let y = Sys.opaque_identity 0 in let f () = y + y + ... + y`). I would expect that even there it's only a small amount of extra memory for the whole compiler.

Reply:
I've tried an example with a free variable occurring 2000 times, and the difference between trunk and this branch is lost in the noise (both in native and bytecode).

Reply:
> the sharing would be lost when we translate to Cmm

Duh, of course, I had missed this.